feat: implementar test unitarios e integrales

DaniGTeamcubation opened the pull request:
* Se agregan test unitarios para todas las capas del proyecto
* Se agregan test integrales para la capa de persistencia
* Se incluyen modificaciones en el código para resolver problemas encontrados con los test

DaniGTeamcubation committed:
feature: agregar test unitarios para PatientService

- se agregan test unitarios para cubrir distintos casos de uso de los metodos
- se modifica el dockerfile borrando el skip de test
- se crea nuevo perfil de test

DaniGTeamcubation committed:
refactor: implementar nueva interfaz para uso de factory en servicio

- Se crea una nueva interfaz ExporterFactoryProviderPortOut
- Se modifican nombres de clases para mantener consistencia

DaniGTeamcubation committed:
fix: email de pacientes no se guarda al crear paciente nuevo

- Se agrega el set de email en el metodo toDomain de PatientRestMapper

DaniGTeamcubation committed:
feature: completar test unitarios de exportación de archivos en servicio

DaniGTeamcubation committed:
fix: reemplazar anotacion @SpringBootTest por @ExtendWith()

- Modifico anotacion en PatientServiceTest para evitar cargar todo el contexto de spring

DaniGTeamcubation committed:
refactor: Mejoras en implementacion de test unitarios

- Se utiliza un builder para generar pacientes necesarios para las pruebas

DaniGTeamcubation committed:
feature: Implementar test unitarios para endpoints del controlador

- Se agrega nueva dependencia para poder validar campos desde el controlador
- Las anotaciones de validacion se agregaron en la interfaz

DaniGTeamcubation committed:
feature: agregar validaciones faltantes a controlador

- se agregan anotaciones faltantes a interfaz para validar datos de entrada

DaniGTeamcubation committed:
feature: agregar tests integrales para metodos de repositorio

- Se crearon dos clases test una con test integrales utilizando h2 otra para usar mock en casos de llamadas a sp
- Se agregaron algunas excepciones GlobalExceptionHandler

DaniGTeamcubation committed:
update: se modifica codigo de excepcion testeado para caso dni duplicado al actualizar paciente

DaniGTeamcubation committed:
refactor: se agregan mejoras a partir de analisis con SonarQube

- Unifica algunos test repetitivos utilizando @ParameterizedTest

DaniGTeamcubation committed:
update: agregar test para factory exportador de formatos

- Se agregan test unitarios para las clases factory

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/application/service/PatientServiceTest.java:
```diff
@@ -0,0 +1,741 @@
+        this.patientService = new PatientService(this.patientRepository, exporterFactoryProvider);
+    }
+
+    //Se puede usar para utilizar paciente por defecto o crear nuevos pacientes
```
Eliminar comentario o usar notacion javadoc

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/application/service/PatientServiceTest.java:
```diff
@@ -0,0 +1,741 @@
+    private PatientService patientService;
+
+    @Mock
+    private PatientRepository patientRepository;
```
Usar el portout en vez del repository

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/application/service/PatientServiceTest.java:
```diff
@@ -0,0 +1,741 @@
+    }
+
+    //Se puede usar para utilizar paciente por defecto o crear nuevos pacientes
+    static class PatientBuilder {
```
Cambiar nombre por TestPatientBuilder y hacer privado y estático para que solo se use en el test

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/application/service/PatientServiceTest.java:
```diff
@@ -0,0 +1,741 @@
+        );
+    }
+
+    //TEST CREAR PACIENTE
```
Eliminar comentarios, reemplazar por anotación @Nested en caso de querer agruparlos por caso de uso

DaniGTeamcubation replied:
Pasé a usar @nested, me encanta

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/factory/CsvExporterFactoryTest.java:
```diff
@@ -0,0 +1,20 @@
+        PatientExporterPortOut exporter = factory.createPatientExporter();
+
+        assertNotNull(exporter);
+        assertTrue(exporter instanceof CsvPatientExporter);
```
Usar assertInstanceOf

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/factory/CsvExporterFactoryTest.java:
```diff
@@ -0,0 +1,20 @@
+import com.teamcubation.api.pacientes.infrastructure.adapter.in.exporter.impl.CsvPatientExporter;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
No usar wildcards, menos por tan pocas anotaciones a importar

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/factory/ExporterFactoryProviderTest.java:
Aplicar mismas observaciones que en los demás exporters

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/factory/JsonExporterFactoryTest.java:
Aplicar mismas observaciones que en los demás exporters

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/impl/CsvPatientExporterTest.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Test
+    void export_shouldReturnCsvString() {
+        CsvPatientExporter exporter = new CsvPatientExporter();
```
Instanciar a nivel clase y usar @BeforeEach

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/impl/CsvPatientExporterTest.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Test
+    void export_emptyList_shouldReturnOnlyHeader() {
+        CsvPatientExporter exporter = new CsvPatientExporter();
```
Instanciar a nivel clase y usar @BeforeEach

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/rest/controller/PatientControllerTest.java:
```diff
@@ -0,0 +1,1016 @@
+    private MockMvc mockMvc;
+    @MockBean
+    private PatientService patientService;
+    @InjectMocks
```
Al usar @WebMvcTest no hace falta anotar el controller con @InjectMocks

gaston-areco-tq commented on src/main/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/exporter/impl/CsvPatientExporter.java:
```diff
@@ -23,6 +23,7 @@ public String export(List<Patient> patients) {
                         patient.getPhoneNumber()))
                 .collect(Collectors.joining("\n"));
 
+        System.out.println(header + body);
```
Eliminar sout

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/rest/controller/PatientControllerTest.java:
```diff
@@ -0,0 +1,1016 @@
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
+        this.objectMapper = new ObjectMapper();
+        this.baseURL = "/v1/pacientes";
```
Usar una constante para este valor

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/in/rest/controller/PatientControllerTest.java:
```diff
@@ -0,0 +1,1016 @@
+        this.baseURL = "/v1/pacientes";
+    }
+
+    static class PatientBuilder {
```
Seguir las observaciones en el PatientBuilder del PatienceServiceTest e intentar abstraer este builder en alguna clase utilitaria para los tests para evitar tanta repetición

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryIT.java:
```diff
@@ -0,0 +1,387 @@
+    private IPatientRepository jpaRepository;
+
+    @Mock
+    private IPatientRepository mockedJpaRepository;
```
Este mock no se usa

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryIT.java:
```diff
@@ -0,0 +1,387 @@
+        this.patientRepository = new PatientRepository(jpaRepository);
+    }
+
+    static class PatientBuilder {
```
Mismo caso que con los demás builders

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryIT.java:
```diff
@@ -0,0 +1,387 @@
+    void setUp() {
+        this.patientRepository = new PatientRepository(jpaRepository);
+    }
+
```
Agregar un @AfterEach para limpiar el repository 

@AfterEach
    void cleanUp() {
        jpaRepository.deleteAll();
    }

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryTest.java:
```diff
@@ -0,0 +1,234 @@
+        List<Patient> result = this.patientRepository.findByHealthInsuranceProvider(provider, limit, offset);
+
+        assertTrue(result.isEmpty());
+    }
```
Agregar verify

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryTest.java:
```diff
@@ -0,0 +1,234 @@
+
+        assertEquals(1, result.size());
+        assertEquals("Lucía", result.get(0).getName());
+    }
```
Agregar verify

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryTest.java:
```diff
@@ -0,0 +1,234 @@
+        List<Patient> result = this.patientRepository.findByHealthInsuranceProvider(null, 5, 0);
+
+        assertTrue(result.isEmpty());
+    }
```
Agregar verify

gaston-areco-tq commented on src/test/java/com/teamcubation/api/pacientes/infrastructure/adapter/out/persistence/repository/PatientRepositoryTest.java:
```diff
@@ -0,0 +1,234 @@
+        List<Patient> result = this.patientRepository.findByHealthInsuranceProvider("", 5, 0);
+
+        assertTrue(result.isEmpty());
+    }
```
Agregar verify

DaniGTeamcubation committed:
update: mejorar PatientServiceTest a partir de comentarios en PR #47

DaniGTeamcubation committed:
update: mejorar clases factory de exportacion a partir de comentarios en PR #47

DaniGTeamcubation committed:
update: mejoras pendientes a partir de comentarios en PR #47

- Se agrega una clase TestPatientBuilder para reutilizar codigo en PatientServiceTest y PatientControllerTest

DaniGTeamcubation committed:
fix: evitar error de Hibernate al limpiar la base tras excepción en tests

- Se eliminó el método @AfterEach con deleteAll() porque @DataJpaTest ya realiza rollback automático para evitar error "null id in entry (don't flush the Session after an exception occurs)"
cuando un test fallaba durante la persistencia.

DaniGTeamcubation committed:
update: mejoras a partir de analisis con sonarQube
